File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Ros/Service/RosServiceResponse.swift
//
//  RosServiceResponse.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/11/15.
//

import Foundation

struct RosServiceResponse<T: RosServiceResponseValuesProtocol>: RosServiceResponseProtocol {
    typealias V = T

    let op: RosBridgeMessageOperation
    let id: String?
    let service: String
    let values: [T]
    let result: Bool

    init(
        op: RosBridgeMessageOperation = .serviceResponse,
        id: String? = nil,
        service: String,
        values: [T],
        result: Bool
    ) {
        self.op = op
        self.id = id
        self.service = service
        self.values = values
        self.result = result
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/ros/service/RosServiceResponse.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.ros.service

import com.shusuke.raspberry_pi_client.data.infrastructure.rosbridge.RosBridgeMessageOperation
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * [Service Response](https://github.com/RobotWebTools/rosbridge_suite/blob/ros2/ROSBRIDGE_PROTOCOL.md)
 *
 * @param T レスポンス値型（EmptyService, TurtlesimServiceResponse 等）
 */
@Serializable
data class RosServiceResponse<T>(
    val op: RosBridgeMessageOperation = RosBridgeMessageOperation.serviceResponse,
    val id: String? = null,
    val service: String,
    val result: Boolean = true,
    @SerialName("values")
    val values: T? = null
)
